### PR TITLE
Full PSP Support

### PR DESCRIPTION
**Pull request by Beyley**
Creating PR as draft for now to perhaps collect potential feedback, still contains a couple hacks and test code, and still has 3 completely unimplemented endpoints, but the tree is *mostly* in an okay state

### CODE REVIEW

File: Refresh.GameServer/Endpoints/Game/Levels/PublishEndpoints.cs [hunk 1]
```diff
@@ -80,8 +80,12 @@ public Response PublishLevel(RequestContext context, GameUser user, Token token,
         level.MinPlayers = Math.Clamp(level.MinPlayers, 1, 4);
         level.MaxPlayers = Math.Clamp(level.MaxPlayers, 1, 4);
 
+        bool pspConnection = context.RequestHeaders.Get("User-Agent") == "LBPPSP CLIENT";
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Levels/PublishEndpoints.cs:83]: this sholkd be an extension of `RequestContext` similar to `GetPageData`

File: Refresh.GameServer/Endpoints/Game/ResourceEndpoints.cs [hunk 2]
```diff
@@ -26,25 +26,31 @@ public class ResourceEndpoints : EndpointGroup
     public Response UploadAsset(RequestContext context, string hash, string type, byte[] body, IDataStore dataStore,
         GameDatabaseContext database, GameUser user, AssetImporter importer, GameServerConfig config, IDateTimeProvider timeProvider)
     {
-        if (dataStore.ExistsInStore(hash))
+        //Check whether the request comes from a PSP or not
+        bool pspConnection = context.RequestHeaders.Get("User-Agent") == "LBPPSP CLIENT";
+
+        string path = pspConnection ? $"psp/{hash}" : hash;
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/ResourceEndpoints.cs:32]:
```suggestion
        string assetPath = pspConnection ? $"psp/{hash}" : hash;
```

**Review by jvyden — CHANGES_REQUESTED** [review]

**Comment by Beyley** [comment]
Commit was reverted as it seems to regress both PPSSPP and real hardware, not sure what the problem is with it

**Comment by jvyden** [comment]
That's odd. Can you tell which endpoint in particular it is, if any?

**Comment by Beyley** [comment]
> That's odd. Can you tell which endpoint in particular it is, if any?

Submit developer (POST to `scoreboard/developer/{id}`) score from my testing, maybe it doesnt want a response on PSP but does on LBP1? i based part of this code from lighthouse, and it returns the scores there

**Comment by Beyley** [comment]
With the latest commit, real hardware is now tested fully working, ready for review!

**Comment by logantgt** [comment]
yeyeye looks good mayte i think

File: Refresh.GameServer/Endpoints/Game/DataTypes/Request/GameLevelRequest.cs [hunk 3]
```diff
@@ -43,7 +43,9 @@ public class GameLevelRequest
     [XmlElement("initiallyLocked")] public bool IsLocked { get; set; }
     [XmlElement("isSubLevel")] public bool IsSubLevel { get; set; }
     [XmlElement("shareable")] public int IsCopyable { get; set; }
-
+    
+    [XmlElement("backgroundGUID")] public string? BackgroundGuid { get; set; }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/DataTypes/Request/GameLevelRequest.cs:47]: Do we know what the format of this is? Is it ever actually used? Thinking about ways we could sanitize this/whether we need it at all

File: Refresh.GameServer/Endpoints/Game/Levels/LeaderboardEndpoints.cs [hunk 4]
```diff
@@ -25,6 +25,27 @@ public Response PlayLevel(RequestContext context, GameUser user, GameDatabaseCon
...
+        return new SerializedLeaderboardResponse(new List<SerializedPlayerLeaderboardResponse>());
+    }
+
+    [GameEndpoint("scoreboard/developer/{id}", ContentType.Xml, Method.Post)]
+    public Response SubmitDeveloperScore(RequestContext context, GameUser user, GameDatabaseContext database, int id, SerializedScore body)
+    {
+        //TODO
+        return new Response(SerializedScoreLeaderboardList.FromSubmittedEnumerable(new List<ScoreWithRank>()), ContentType.Xml);
+    }
+
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Levels/LeaderboardEndpoints.cs:48]: I think these should definitely be implemented, at least for user levels. Also `scoreboard/user/{id}` might be a duplicate? This seems a little strange.

File: Refresh.GameServer/Endpoints/Game/ResourceEndpoints.cs [hunk 5]
```diff
@@ -26,25 +27,31 @@ public class ResourceEndpoints : EndpointGroup
     public Response UploadAsset(RequestContext context, string hash, string type, byte[] body, IDataStore dataStore,
         GameDatabaseContext database, GameUser user, AssetImporter importer, GameServerConfig config, IDateTimeProvider timeProvider)
     {
-        if (dataStore.ExistsInStore(hash))
+        //Check whether the request comes from a PSP or not
+        bool pspConnection = context.IsPSP();
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/ResourceEndpoints.cs:31]: Should be named `isPSP` for consistency

File: Refresh.GameServer/Endpoints/Game/ReviewEndpoints.cs [hunk 6]
```diff
@@ -23,4 +23,46 @@ public Response SubmitRating(RequestContext context, GameDatabaseContext databas
...
+    [GameEndpoint("rate/user/{id}", ContentType.Xml, Method.Post)]
+    [AllowEmptyBody]
+    public Response RateUserLevel(RequestContext context, GameDatabaseContext database, GameUser user, int id)
+    {
+        string? ratingString = context.QueryString.Get("rating");
+        
+        if (ratingString == null)
+        {
+            return BadRequest;
+        }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/ReviewEndpoints.cs:36]:
```suggestion
        if (ratingString == null) return BadRequest;
```

File: Refresh.GameServer/Extensions/RequestContextExtensions.cs [hunk 7]
```diff
@@ -18,4 +18,10 @@ public static (int, int) GetPageData(this RequestContext context, bool api = fal
 
         return (skip, count);
     }
+
+    [Pure]
+    public static bool IsPSP(this RequestContext context)
+    {
+        return context.RequestHeaders.Get("User-Agent") == "LBPPSP CLIENT";
+    }
```
Comment (jvyden) [inline Refresh.GameServer/Extensions/RequestContextExtensions.cs:26]: In general we tend to use arrow methods for small helper functions:

```suggestion
    [Pure]
    public static bool IsPSP(this RequestContext context) => context.RequestHeaders.Get("User-Agent") == "LBPPSP CLIENT";
```

File: Refresh.GameServer/Types/Assets/GameAssetType.cs [hunk 8]
```diff
@@ -98,4 +98,8 @@ public enum GameAssetType
     /// Magic: PTGb
     /// </remarks>
     Painting = 12,
+    /// <summary>
+    /// An image, stored in a TGA container.
+    /// </summary>
```
Comment (jvyden) [inline Refresh.GameServer/Types/Assets/GameAssetType.cs:103]: Should leave remarks about how there technically is no magic, add a seealso to point to the function that detects it

File: Refresh.GameServer/Types/Levels/GameMinimalLevelResponse.cs [hunk 9]
```diff
@@ -82,6 +86,7 @@ public class GameMinimalLevelResponse : IDataConvertableFrom<GameMinimalLevelRes
             MinPlayers = level.MinPlayers,
             TotalPlayCount = level.TotalPlayCount,
             UniquePlayCount = level.UniquePlayCount,
+            AverageRating = 3, //TODO
```
Comment (jvyden) [inline Refresh.GameServer/Types/Levels/GameMinimalLevelResponse.cs:89]: 0 is a better default imo, I think it shows that there are no ratings to the user in a simpler way
```suggestion
            AverageRating = 0, //TODO
```

File: RefreshTests.GameServer/Tests/Middlewares/DigestMiddlewareTests.cs [hunk 10]
```diff
@@ -49,8 +49,8 @@ public async Task DigestIsCorrect()
         using MemoryStream blankMs = new();
         using MemoryStream expectedResultMs = new(Encoding.ASCII.GetBytes(expectedResultStr));
         
-        string serverDigest = DigestMiddleware.CalculateDigest(endpoint, expectedResultMs, "");
-        string clientDigest = DigestMiddleware.CalculateDigest(endpoint, blankMs, "");
+        string serverDigest = DigestMiddleware.CalculateDigest(endpoint, expectedResultMs, "", null, null);
+        string clientDigest = DigestMiddleware.CalculateDigest(endpoint, blankMs, "", null, null);
```
Comment (jvyden) [inline RefreshTests.GameServer/Tests/Middlewares/DigestMiddlewareTests.cs:53]: We should also test the new PSP logic

File: Refresh.GameServer/Importing/Importer.cs [hunk 11]
```diff
@@ -59,6 +59,41 @@ private static bool MatchesMagic(ReadOnlySpan<byte> data, ulong magic)
         BitConverter.TryWriteBytes(magicSpan, BinaryPrimitives.ReverseEndianness(magic));
         return MatchesMagic(data, magicSpan);
     }
+
+    /// <summary>
+    /// Tries to detect TGA files sent from the PSP
+    /// </summary>
+    /// <param name="data"></param>
+    /// <returns></returns>
+    private static bool IsPspTga(ReadOnlySpan<byte> data)
```
Comment (jvyden) [inline Refresh.GameServer/Importing/Importer.cs:68]: Another safety check we could add is potentially bringing in the GameToken to check it's version & platform to see if it matches with PSP.

Mainline & Vita will never upload TGA files.

**Review by jvyden — COMMENTED** [review]
Looks good! There are just a couple things we have to consider for PSP as we bring in its quirks.

File: Refresh.GameServer/Endpoints/Game/Levels/LeaderboardEndpoints.cs (hunk 4 above)
Comment (jvyden) [reply Refresh.GameServer/Endpoints/Game/Levels/LeaderboardEndpoints.cs:48]: Oops, I was confusing this for topscores/user/id/{type}. My bad.

File: Refresh.GameServer/Endpoints/Game/DataTypes/Request/GameLevelRequest.cs (hunk 3 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/Game/DataTypes/Request/GameLevelRequest.cs:47]: I believe it is just a `GUID` asset ID (`gXXXXX`), however i dont believe ive actually seen the game upload levels with this `backgroundGUID` field set yet, i just saw it when RE and figured "might as well send it back to make the game happy", i'll do a deeper dive later

Comment (jvyden) [reply Refresh.GameServer/Endpoints/Game/DataTypes/Request/GameLevelRequest.cs:47]: I believe mainline can sometimes send this too, but I haven't nailed down a pattern on when/where. We can address this after merging.

**Review by jvyden — APPROVED** [review]
PSP IS REAL

**Comment by speedyman09** [comment]
WHA

**Pull request merged**